components/script/dom/gpubuffer.rs

```diff
@@ -104,19 +127,68 @@ impl Drop for GPUBuffer {
 }
 
 impl GPUBufferMethods for GPUBuffer {
+    #[allow(unsafe_code)]
     /// https://gpuweb.github.io/gpuweb/#dom-gpubuffer-unmap
     fn Unmap(&self) {
-        self.channel
-            .0
-            .send(WebGPURequest::UnmapBuffer(self.buffer))
-            .unwrap();
+        let cx = self.global().get_cx();
+        // Step 1
+        match *self.state.borrow() {
+            GPUBufferState::Unmapped | GPUBufferState::Destroyed => {
+                // TODO: Record validation error on the current scope
+                return;
+            },
+            GPUBufferState::MappedForWriting => {
+                // Step 3.1
+                match ArrayBuffer::from(self.mapping.get()) {
+                    Ok(array_buffer) => {
+                        self.channel
+                            .0
+                            .send(WebGPURequest::UnmapBuffer(
+                                self.device.0,
+                                self.id(),
+                                array_buffer.to_vec(),
+                            ))
+                            .unwrap();
+                        // Step 3.2
+                        unsafe {
+                            DetachArrayBuffer(
+                                *cx,
+                                Handle::from_marked_location(array_buffer.underlying_object()),
+                            );
+                        }
+                    },
+                    _ => {
+                        // Step 2
+                        unsafe {
+                            let handle = Handle::from_marked_location(self.mapping.get() as _);
```

Comment:
I think we can't use it because `self.mapping.handle()` returns a `js::rust::Handle` but `IsPromiseObject`, `RejectPromise` and `DetachArrayBuffer` accepts `js::jsapi::Handle`

Reply:
Import IsPromiseObject and DetachArrayBuffer from js::rust::wrapped instead of jsapi. Those version accept rust::Handle. Alternativey use .into() to convert rust::Handle to jsapi::Handle.

Reply:
Thank you! Importing the right one is did the trick!